crates/guest-rust/src/rt/async_support/inter_task_wakeup.rs

```diff
@@ -0,0 +1,97 @@
+use super::FutureState;
+use crate::rt::async_support::{BLOCKED, COMPLETED};
+use crate::{RawStreamReader, RawStreamWriter, StreamOps, UnitStreamOps};
+use std::sync::Mutex;
+
+#[derive(Default)]
+pub struct State {
+    /// A lazily-initialized stream used to signal inter-task notifications.
+    ///
+    /// This stream is used when one component-model task is used to wake up
+    /// another component-model task. This can happen when the async event being
+    /// waited on is defined purely in Rust, for example, and doesn't rely on
+    /// any component-model primitives.
+    stream: Option<RawStreamReader<UnitStreamOps>>,
+    /// Boolean if there's an active read of `inter_task_stream`. Used to handle
+    /// cancellation/deduplication of the read.
+    stream_reading: bool,
+}
+
+impl FutureState<'_> {
+    pub(super) fn read_inter_task_stream(&mut self) {
+        // Lazily allocate the inter-task stream now that we're actually going
+        // to sleep. We don't know where the wakeup notification will come from
+        // so it's required to allocate one here.
+        if self.inter_task_wakeup.stream.is_none() {
+            assert!(!self.inter_task_wakeup.stream_reading);
+            let (writer, reader) = UnitStreamOps::new();
+            self.inter_task_wakeup.stream = Some(reader);
+            let mut waker_stream = self.waker.inter_task_stream.lock.lock().unwrap();
+            assert!(waker_stream.is_none());
+            *waker_stream = Some(writer);
+        }
+
+        // If there's not already a pending read then schedule a new read here.
+        //
+        // Note that this should always return `BLOCKED` since as the only task
+        // running it's not possible for a read to be anywhere else in the
+        // system. Additionally we keep the read end alive, so this shouldn't
+        // ever returned dropped/closed either.
+        if !self.inter_task_wakeup.stream_reading {
+            let stream = self.inter_task_wakeup.stream.as_mut().unwrap();
+            let handle = stream.handle();
+            let rc = unsafe { UnitStreamOps.start_read(handle, 1 as *mut _, 1) };
```

Comment:
What's the reason for using `1 as *mut _` here (and below) vs. `ptr::null_mut()`?